File: src/main/services/pdfPath.ts
import { isAbsolute, resolve as resolvePath } from 'node:path'
import { MainProcessError } from './errors'
import { ExistingPathError, resolveExistingPath } from './existingPath'

export function resolvePdfFilePath(rawPath: string): string {
  if (!rawPath || !isAbsolute(rawPath)) {
    throw new MainProcessError('invalid_path', 'PDF path must be absolute')
  }
  const resolved = resolvePath(rawPath)
  if (!resolved.toLowerCase().endsWith('.pdf')) {
    throw new MainProcessError('invalid_path', 'Selected file must be a PDF')
  }
  try {
    return resolveExistingPath(resolved, 'file')
  } catch (error) {
    if (error instanceof ExistingPathError && error.failure === 'missing') {
      throw new MainProcessError('file_missing', `File not found: ${resolved}`)
    }
    if (
      error instanceof ExistingPathError &&
      (error.failure === 'symbolic_link' || error.failure === 'wrong_kind')
    ) {
      throw new MainProcessError('invalid_path', 'Selected path must be a regular PDF file')
    }
    throw new MainProcessError('invalid_path', `Unable to inspect PDF file: ${resolved}`)
  }
}

File: backend/refora_server/library/pdf_path.py
from __future__ import annotations

import os

from refora_server.db.errors import RepoError


def resolvePdfFilePath(rawPath: str) -> str:
    if not rawPath or not os.path.isabs(rawPath):
        raise RepoError("invalid_path", "PDF path must be absolute")
    resolved = os.path.normpath(os.path.abspath(rawPath))
    if not resolved.lower().endswith(".pdf"):
        raise RepoError("invalid_path", "Selected file must be a PDF")
    if not os.path.exists(resolved):
        raise RepoError("file_missing", f"File not found: {resolved}")
    try:
        if os.path.islink(resolved) or not os.path.isfile(resolved):
            raise RepoError("invalid_path", "Selected path must be a regular PDF file")
    except RepoError:
        raise
    except Exception:
        raise RepoError("invalid_path", f"Unable to inspect PDF file: {resolved}")
    return resolved
